pkg/tunnel/connection/splice.go

```diff
@@ -35,55 +43,130 @@ func Splice(tunDev tun.Device, conn Connection) error {
-					continue
+					return fmt.Errorf("failed to write to connection: %w", err)
 				}
 				if len(icmp) > 0 {
 					slog.Debug("Sending ICMP packet")
 					if _, err := tunDev.Write([][]byte{icmp}, 0); err != nil {
+						if strings.Contains(err.Error(), "closed") {
+							slog.Debug("TUN device closed")
+							return net.ErrClosed
+						}
+
 						slog.Error("Failed to write ICMP packet", slog.Any("error", err))
+						return fmt.Errorf("failed to write ICMP packet: %w", err)
 					}
 				}
 			}
 		}
 	})
 
 	g.Go(func() error {
-		pkts := make([][]byte, batchSize)
-		for i := range pkts {
-			pkts[i] = make([]byte, netstack.IPv6MinMTU+tunOffset)
+		defer func() {
+			slog.Debug("Stopped reading from connection")
+		}()
+
+		var pktPool = sync.Pool{
+			New: func() any {
+				return ptr.To(make([]byte, netstack.IPv6MinMTU+tunOffset))
+			},
 		}
 
-		// TODO: batched write to TUN device, unfortunately ReadPacket() is blocking
-		// and not batched which makes this tricky.
+		pktCh := make(chan *[]byte, batchSize)
 
-		for {
-			n, err := conn.ReadPacket(pkts[0][tunOffset:])
-			if err != nil {
-				return fmt.Errorf("failed to read from connection: %w", err)
+		g.Go(func() error {
+			defer close(pktCh)
+
+			for {
+				pkt := pktPool.Get().(*[]byte)
+				n, err := conn.ReadPacket((*pkt)[tunOffset:])
+				if err != nil {
+					if strings.Contains(err.Error(), "closed") {
+						slog.Debug("Connection closed")
+						return net.ErrClosed
+					}
+
+					slog.Error("Failed to read from connection", slog.Any("error", err))
+					return fmt.Errorf("failed to read from connection: %w", err)
+				}
+
+				slog.Debug("Read packet from connection", slog.Int("len", n))
+
+				*pkt = (*pkt)[:n+tunOffset]
+				pktCh <- pkt
 			}
+		})
+
+		pkts := make([][]byte, batchSize)
+
+		for {
+			select {
+			case pkt, ok := <-pktCh:
+				if !ok {
+					return nil
+				}
+
+				pkts[0] = *pkt
+				batchCount := 1
 
-			slog.Debug("Read from connection", slog.Int("bytes", n))
+				closed := false
+			gatherBatch:
+				for batchCount < batchSize && !closed {
+					select {
+					case pkt, ok := <-pktCh:
+						if !ok {
+							closed = true
+							break
+						}
+						pkts[batchCount] = *pkt
+						batchCount++
+					default:
+						break gatherBatch
+					}
+				}
+
+				stats.recordWriteBatch(batchCount)
+
+				if _, err := tunDev.Write(pkts[:batchCount], tunOffset); err != nil {
+					if strings.Contains(err.Error(), "closed") {
+						slog.Debug("TUN device closed")
+						return net.ErrClosed
+					}
 
-			if _, err := tunDev.Write([][]byte{pkts[0][:n+tunOffset]}, tunOffset); err != nil {
-				slog.Error("Failed to write to TUN", slog.Any("error", err))
-				continue
+					slog.Error("Failed to write to TUN", slog.Any("error", err))
+					return fmt.Errorf("failed to write to TUN: %w", err)
+				}
+
+				for i := 0; i < batchCount; i++ {
+					pkt := pkts[i][:cap(pkts[i])]
+					pktPool.Put(&pkt)
```

Comment:
```suggestion
					pkt := pkts[i][:cap(pkts[i])]
					pkt = pkt[:]
					pktPool.Put(&pkt)
```
I think we need to reset size here or next time `n, err := conn.ReadPacket((*pkt)[tunOffset:])` can be cut short

Reply:
actually nvm i see line 167 already doing this